File: example_py/basics7_display_joint_commands.py
#!/usr/bin/env python

"""basics1_display_joint_commands.py

This tutorial will check connection with the robot server and print joint commands.
"""

__copyright__ = "Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import time
import argparse
import threading
import spdlog  # pip install spdlog

# Import Flexiv DDK Python library
# fmt: off
import sys
sys.path.insert(0, "../lib_py")
import flexivddk
# fmt: on



def print_joint_commands(client, logger, stop_event):
    """
    Print joint commands data @ 1Hz.

    """

    while not stop_event.is_set():
        # Print joint commands, round all float values to 2 decimals
        logger.info("Current robot joint commands:")
        # fmt: off
        print("{")
        print(f"q: {['%.2f' % i for i in client.joint_commands().q_des]}",)
        print(f"dq_des: {['%.2f' % i for i in client.joint_commands().dq_des]}")
        print(f"tau_des: {['%.2f' % i for i in client.joint_commands().tau_des]}")
        print("}", flush= True)
        # fmt: on

        time.sleep(1)


def main():
    # Create an event to signal the thread to stop
    stop_event = threading.Event()

    # Program Setup
    # ==============================================================================================
    # Parse arguments
    argparser = argparse.ArgumentParser()
    argparser.add_argument(
        "robot_sn",
        help="Serial number of the robot to connect. Remove any space, e.g. Rizon4s-123456",
    )
    args = argparser.parse_args()

    # Define alias
    logger = spdlog.ConsoleLogger("Example")

    # Print description
    logger.info(
        ">>> Tutorial description <<<\nThis tutorial will check connection "
        "with the robot server and print robot joint commands.\n"
    )

    try:
        # Instantiate client interface
        # ==========================================================================================
        client = flexivddk.Client(args.robot_sn)

        if not client.connected():
            logger.warn("Cannot get connected with robot, retrying ...")
            if not client.connected()():
                logger.error("Exiting ...")
                return 1
        logger.info(f"Connected with robot {args.robot_sn}")
    except Exception as e:
        # Print exception error message
        logger.error(str(e))

    # Thread for printing robot data
    # =============================================================================
    print_thread = threading.Thread(
        target=print_joint_commands, args=[client, logger, stop_event]
    )
    print_thread.start()

    # Use main thread to catch keyboard interrupt and exit thread
    try:
        while not stop_event.is_set():
            time.sleep(0.1)
    except KeyboardInterrupt:
        # Send signal to exit thread
        logger.info("Stopping print thread")
        stop_event.set()

    # Wait for thread to exit
    print_thread.join()
    logger.info("Print thread exited")


if __name__ == "__main__":
    main()

File: example/basics7_display_joint_commands.cpp
/**
 * @example basics7_display_joint_commands.cpp
 * This tutorial check connection with the robot and print received robot joint
 * commands.
 * @copyright Copyright (C) 2016-2024 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */
#include <atomic>
#include <csignal>
#include <flexiv/ddk/client.hpp>
#include <flexiv/ddk/utility.hpp>
#include <iostream>
#include <spdlog/spdlog.h>
#include <thread>

namespace {
/** Atomic signal to stop periodic print tasks */
std::atomic<bool> keep_running(true);
} // namespace

/**
 * @brief Handler for user stop interaction
 */
void SignalHandler(int signal)
{
    if (signal == SIGINT) {
        keep_running = false;
        spdlog::info("Signal received, stopping...");
    }
}

/** @brief Print program usage help */
void PrintHelp()
{
    // clang-format off
    std::cout << "Required arguments: [robot SN]" << std::endl;
    std::cout << "    robot SN: Serial number of the robot to connect to. "
                 "Remove any space, for example: Rizon4s-123456" << std::endl;
    std::cout << "Optional arguments: None" << std::endl;
    std::cout << std::endl;
    // clang-format on
}

/** @brief Print robot joint commands data @ 1Hz */
void printJointCommands(flexiv::ddk::Client& client)
{
    while (keep_running.load()) {
        // Check connection with the robot
        if (!client.connected()) {
            spdlog::error("Cannot connect with robot, retrying...");
            std::this_thread::sleep_for(std::chrono::seconds(5));
            continue;
        }
        // Print all joint commands in JSON format using the built-in ostream
        // operator overloading
        spdlog::info("Current robot joint commands:");
        std::cout << client.joint_commands() << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

int main(int argc, char* argv[])
{
    // Program Setup
    if (argc < 2 || flexiv::ddk::utility::ProgramArgsExistAny(argc, argv, {"-h", "--help"})) {
        PrintHelp();
        return 1;
    }
    std::string robot_sn = argv[1];

    // Print description
    spdlog::info(
        ">>> Tutorial description <<<\nThis tutorial check connection "
        "with the robot and print received robot joint commands.");

    // Setup signal handler for graceful exit
    std::signal(SIGINT, SignalHandler);

    try {
        // DDK Initialization
        flexiv::ddk::Client client(robot_sn);

        // Print Commands
        // =========================================================================================
        // Use std::thread to do scheduling so that this example can run on all OS
        std::thread low_priority_thread(std::bind(printJointCommands, std::ref(client)));

        // Properly exit thread
        low_priority_thread.join();

    } catch (const std::exception& e) {
        spdlog::error(e.what());
        return 1;
    }

    return 0;
}
